<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>AmBre by l337x911</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">AmBre</h1>
        <p class="header">Amplification of breakpoints for reliable genome structural variation detection.</p>

        <ul>
          <li class="download"><a class="buttons" href="https://github.com/l337x911/AmBre/blob/master/dist/AmBre-1.0release.zip">Download ZIP</a></li>
          <li class="download"><a class="buttons" href="https://github.com/l337x911/AmBre/blob/master/dist/AmBre-1.0release.tar.gz">Download TAR</a></li>
          <li><a class="buttons github" href="https://github.com/l337x911/AmBre">View On GitHub</a></li>
        </ul>

        <p class="header">This project is maintained by <a class="header name" href="https://github.com/l337x911">l337x911</a></p>


      </header>
      <section>
        <h1></h1>

<h1>
<a name="ambre-primer-designing-and-sequence-analysis-for-amplifying-breakpoints" class="anchor" href="#ambre-primer-designing-and-sequence-analysis-for-amplifying-breakpoints"><span class="octicon octicon-link"></span></a>AmBre: Primer designing and sequence analysis for amplifying breakpoints.</h1>

<p>Developed by Anand D. Patel in Vineet Bafna's lab at University of California, San Diego
patel [dot] anandd [at] gmail [dot] com</p>

<p>AmBre is composed of two parts. A) Primer Designing
B) Long read sequence analysis.</p>

<pre><code>ambre_design &lt;reference.fasta&gt; &lt;regions.txt&gt; [&lt;temptag&gt;]
ambre_analyze &lt;reference.fasta&gt; &lt;contig&gt; &lt;mapped_reads.sam&gt; [&lt;temptag&gt;]
</code></pre>

<h1>
<a name="installing" class="anchor" href="#installing"><span class="octicon octicon-link"></span></a>Installing</h1>

<h2>
<a name="software-requirements" class="anchor" href="#software-requirements"><span class="octicon octicon-link"></span></a>Software Requirements</h2>

<p>AmBre requires Python 2.7.x and numpy</p>

<p>For primer designing, the following software needs to be installed</p>

<ul>
<li><p>Primer3 Version 2.3.0 <a href="http://sourceforge.net/projects/primer3/">http://sourceforge.net/projects/primer3/</a></p></li>
<li><p>Blat Version 34x12 <a href="http://hgdownload.cse.ucsc.edu/admin/exe/">http://hgdownload.cse.ucsc.edu/admin/exe/</a></p></li>
<li><p>Multiplx Version 1.2 <a href="http://bioinfo.ut.ee/?page_id=167">http://bioinfo.ut.ee/?page_id=167</a></p></li>
</ul><p>Sequence analysis requires PacBio's BLASR and SMRT pipe.</p>

<ul>
<li>BLASR <a href="https://github.com/PacificBiosciences/blasr">https://github.com/PacificBiosciences/blasr</a>
</li>
<li>PacBio SMRT-Analysis <a href="https://github.com/PacificBiosciences/SMRT-Analysis">https://github.com/PacificBiosciences/SMRT-Analysis</a>
</li>
</ul><h2>
<a name="installing-ambre" class="anchor" href="#installing-ambre"><span class="octicon octicon-link"></span></a>Installing AmBre</h2>

<p>AmBre can be run from source or installed.</p>

<p>Download the tar or zip ball and install using the setup.py script.</p>

<pre><code>    python setup.py install
</code></pre>

<p>To configure the above binary dependencies, edit "ambre.conf" file. The config file
is located with</p>

<pre><code>    ambre_test -c
</code></pre>

<p>Edit the file paths for primer3_dir, aligner_bin, and multiplx to locate the dependencies.
For primer3_dir set the file path to the directory containing primer3_core.
 In the config file specify number of threads and a temp directory to store intermediate files.
 Note, user-specific config files can be used by running any module with the --config argument.</p>

<p>To verify the installation/configuration is correct and
 run a small test case for AmBre-design and AmBre-analyze (takes ~5min). ::</p>

<pre><code>    ambre_test
</code></pre>

<p>If all is well, continue to usage.</p>

<p>Alternatively, to run from source directory use ::</p>

<pre><code>    python -m ambre.test.install_unit -c
    python -m ambre.test.install_unit
</code></pre>

<h1></h1>

<h1>
<a name="ambre-design-usage" class="anchor" href="#ambre-design-usage"><span class="octicon octicon-link"></span></a>AmBre-design Usage</h1>

<h1>
<a name="description" class="anchor" href="#description"><span class="octicon octicon-link"></span></a>Description</h1>

<p>AmBre-design designs an assay to detect structural variations. Given input
regions where one expects breakpoints for a particular structural variation
(SV), AmBre-design finds primers uniformly tiled across the regions to reliably
amplify DNA harboring the structural variation with PCR.</p>

<p>Each structural variation is a rearrangement of reference DNA segments to
produce a novel adjacency in the donor sample. However, it is possible to know
that a SV occurs without knowing the exact boundaries of the rearranged
reference segments. In PCR, a forward primer (DNA sequence) designed to the
left and a reverse primer (reverse complement DNA sequence) designed to the
right of the donor sample adjacency would replicate DNA harboring the SV. PCR
is only capable of reproducing a limited DNA length, so primers must be
designed within a certain distance of the donor adjacency. Using multiple
forward primers and reverse primers across a breakpoint region ensures that
some forward and reverse primer appear within an amplifiable distance around
the donor sample adjacency.</p>

<p>See Patel et al. (submitted) for details.</p>

<h1>
<a name="to-run" class="anchor" href="#to-run"><span class="octicon octicon-link"></span></a>To Run</h1>

<p>Designing requires a reference sequence (fasta format) and a set of forward and
reverse intervals to select primers from. The  is a tab-delimited
file specifies these intervals as rows::</p>

<pre><code>&lt;contig name&gt;\t&lt;interval start&gt;\t&lt;interval length&gt;\t&lt;orientation&gt;
...
...
</code></pre>

<p>where orientation is "forward" for forward or "reverse" if reverse.
See "examples/regions.test" for an example regions file.  If specified, resulting
primer designs are given in .out. </p>

<pre><code>ambre_design &lt;reference.fasta&gt; &lt;regions.txt&gt; [&lt;temptag&gt;]
</code></pre>

<p>or from source::</p>

<pre><code>python -m ambre.design.workflow &lt;reference.fasta&gt; &lt;regions.txt&gt; [&lt;temptag&gt;]
</code></pre>

<p>Parameters are defined in the the config file "ambre.conf". A user-defined
config file can be used with the "--config" argument.</p>

<p>Primer designing involves numerous parameters to ensure primers are viable for
PCR. AmBre requires primers to be compatible with one another and evenly tile
supposed regions. Specifying too large or too many regions requires more primers
and leads to unreliable PCR. See below for fine-tuning parameters.</p>

<p>Temptag is used as an identifier for intermediate files. 
It is recommended to specify temptag as
a directory or directory along with a prefix expected for AmBre-design
intermediate files.</p>

<p>For example, to reprint primer solutions and generate a summary figure for
primer locations on input regions using temptag.</p>

<pre><code>ambre_design -c &lt;reference.fasta&gt; &lt;regions.txt&gt; &lt;temptag&gt;
</code></pre>

<p>or from source::</p>

<pre><code>python -m ambre.design.workflow -c &lt;reference.fasta&gt; &lt;regions.txt&gt; &lt;temptag&gt;
</code></pre>

<p>The primer design output is::</p>

<pre><code>&lt;design cost&gt;\t&lt;time seconds&gt;\t&lt;sa parameters&gt;
&lt;contig&gt;\t&lt;position on contig&gt;\t&lt;is_forward&gt;\t&lt;primer sequence&gt;\t&lt;primer3 score&gt;
...
...
=
</code></pre>

<h1>
<a name="parameters" class="anchor" href="#parameters"><span class="octicon octicon-link"></span></a>Parameters</h1>

<p>In the config file, modifying the following parameters
changes the primer designing task.</p>

<p>The approximate distance between primers.</p>

<pre><code>design_primer_spacing_p=6500
</code></pre>

<p>The density of primers in the candidate primer selection task,
that is (Nd/L-1) where <strong>N</strong> is the number of <em>desired</em> primers
<strong>d</strong> is the spacing between primers and <strong>L</strong> is the total 
spacing. See Patel et al. (submitted) for details. </p>

<pre><code>design_primer_density_p=0.2
</code></pre>

<p>Candidate primer generation is performed with Primer3 (Rozen et al. 2000)
and the primer3 parameters file for long range PCR 
used in Patel et al. (submitted) is listed in the "ambre/data" directory.
Another primer3 parameter file can be used by adding the following
line to the config file.</p>

<pre><code>primer3_long_p=/&lt;file&gt;/&lt;path&gt;/&lt;primer3&gt;/&lt;parameters.txt&gt;
</code></pre>

<h2>
<a name="fine-tune-design-parameters" class="anchor" href="#fine-tune-design-parameters"><span class="octicon octicon-link"></span></a>Fine-tune design parameters</h2>

<p>The number of candidate primers to request from primer3
per kilobase</p>

<pre><code>design_primer3_primers_per_kbp_p=50
</code></pre>

<p>Filtering for primer3 penalty criteria</p>

<pre><code>design_max_primer3_penalty_p=1.2
</code></pre>

<p>Filtering primers with reference alignments within,
    design_max_amp_dist=20000</p>

<h1></h1>

<h1>
<a name="ambre-analyze-usage" class="anchor" href="#ambre-analyze-usage"><span class="octicon octicon-link"></span></a>AmBre-analyze Usage</h1>

<h1>
<a name="description-1" class="anchor" href="#description-1"><span class="octicon octicon-link"></span></a>Description</h1>

<p>Analyzes Pacifc Biosciences amplicon resequencing data where amplicons contain
structural variations (SV). AmBre-analyze clusters reads that support the same
SV and then calls breakpoint and consensus amplicon sequences, despite reads
having high insertion and deletion error rates. </p>

<p>Sequence analysis involves a) local alignment with BLASR, b) alignment segmentation,
 c) geometric SV clustering
approach similar to Sindi et al. (2009). d) breakpoint and consensus sequence 
refinement. 
See Patel et al. (submitted) for details.</p>

<h1>
<a name="to-run-" class="anchor" href="#to-run-"><span class="octicon octicon-link"></span></a>To Run </h1>

<p>AmBre-analyze requires a local alignment file (hard-clipped SAM
format) with the corresponding reference file (fasta format) and contig name.
Current implementation only supports intra-contig SVs. Estimated breakpoints
are reported in ".bp" and estimated amplicon sequences are reported
in ".fasta".</p>

<p>Given a "aligned_reads.bas.h5" file from PacBio.</p>

<pre><code>/blasr/alignment/bin/blasr aligned_reads.bas.h5 &lt;reference.fasta&gt; -clipping hard -sam -out &lt;aligned_reads_h5.sam&gt;

ambre_analyze &lt;reference.fasta&gt; &lt;contig&gt; &lt;aligned_reads.sam&gt; [&lt;temptag&gt;]
</code></pre>

<p>or from source::</p>

<pre><code>python -m ambre.analyze.workflow &lt;reference.fasta&gt; &lt;contig&gt; &lt;aligned_reads.sam&gt; [&lt;temptag&gt;]
</code></pre>

<p>The output ".bp" has each breakpoint as an entry with the following form ::</p>

<pre><code>#&lt;name&gt;\t\t\t&lt;mode left bp&gt;\t&lt;mode right bp&gt;
&lt;read_idx&gt;\t&lt;bp on frag idx&gt;\t&lt;left bp&gt;\t&lt;right bp&gt;\t&lt;d&gt;
...
...
=
</code></pre>

<p>The next step is to perform amplicon refinement using PacBio Amplicon
Resequencing Protocol.  Using SMRT-Analysis 2.1, call consensus sequencing
using the estimated amplicon ".fasta" as the reference and the entire
read set.</p>

<p>Parameters are defined in the the config file "ambre.conf". A user-defined
config file can be used with the "--config" argument. See below for fine-tuning
parameters.</p>

<p>Temptag is used as an identifier for intermediate files.  It is recommended to
specify temptag as a directory or directory along with a prefix expected for
AmBre-analyze (can be the same prefix as AmBre-design) intermediate files. </p>

<h2>
<a name="dealing-with-amplicons-with-multiple-breaks" class="anchor" href="#dealing-with-amplicons-with-multiple-breaks"><span class="octicon octicon-link"></span></a>Dealing with amplicons with multiple breaks</h2>

<p>If independent breakpoints found in the previous section in AmBre-analyze
belong to the same amplicon, then annotate the breakpoint headers in the
".bp" file where instead of "#\t" is "#A549_01\t", "#A549_02", and
"..." to represent ordering of breaks along an amplicon. </p>

<pre><code>python ambre_analyze.py --multi-break &lt;reference.fasta&gt; &lt;contig&gt; &lt;aligned_reads.sam&gt; [&lt;temptag&gt;]
</code></pre>

<p>The new ".fasta" file will have updated amplicon sequences containing
multiple breaks.</p>

<h1>
<a name="parameters-1" class="anchor" href="#parameters-1"><span class="octicon octicon-link"></span></a>Parameters</h1>

<p>In the config file, modifying the following parameters
changes the sequence analysis task. See Patel et al. (submitted) for details.</p>

<p>Scoring for local alignment segmetnation
and filtering. Mismatch fraction is because CIGAR format only retains aligned bases (M)
 not misaligned.</p>

<pre><code>analyze_mismatch_fraction_p=0.02
analyze_match_score_p=1
analyze_mismatch_penalty_p=-3
analyze_gapopen_penalty_p=-1
analyze_gapext_penalty_p=-0.2
analyze_breakpoint_penalty_p=-50
</code></pre>

<p>Filtering for minimum cluster size</p>

<pre><code>analyze_min_cluster_size_p=25
</code></pre>

<p>More filtering criteria for breakpoints</p>

<pre><code>analyze_min_sv_dist_p=10
analyze_max_d_p=100
analyze_default_segment_p=6500
</code></pre>
      </section>
      <footer>
        <p><small>Hosted on <a href="http://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		          <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-40555418-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
